### PR TITLE
Update for split-out openshift-sdn proxy plugin

### PR DESCRIPTION
**Pull request by dcbw**
See https://github.com/openshift/openshift-sdn/pull/300 for the openshift-sdn side.

Still not entirely standalone but it should be a lot easier now.  This assumes the proxy will also take a --network-plugin argument (or something like that) which the proxy plugin code now interprets.

### CODE REVIEW

**Comment by dcbw** [comment]
@smarterclayton a bit of progress on this front...

**Comment by smarterclayton** [comment]
Good to see - I've been thinking about how to secure endpoints controller (so we can move the filter server side), so we should chat about that soon

File: pkg/cmd/server/kubernetes/node.go [hunk 1]
```diff
@@ -410,6 +410,9 @@ func (c *NodeConfig) RunProxy() {
 	} else {
 		c.FilteringEndpointsHandler.SetBaseEndpointsHandler(endpointsHandler)
 		endpointsConfig.RegisterHandler(c.FilteringEndpointsHandler)
+		if err := c.FilteringEndpointsHandler.Start(); err != nil {
```
Comment (danwinship) [inline pkg/cmd/server/kubernetes/node.go]: Could SetBaseEndpointsHandler() be merged into Start()?

**Comment by pravisankar** [comment]
LGTM

**Comment by smarterclayton** [comment]
Please see https://github.com/openshift/origin/pull/7807 for the changes I'm making to how the service caches in the node are being initialized.

**Comment by dcbw** [comment]
@smarterclayton the DNS stuff from #7807 doesn't seem like it'll conflict much with PR.  Just tested out separate "openshift start node --disable=proxy" and "openshift start network --disable=plugins" and that seems to work as expected.

**Comment by dcbw** [comment]
@smarterclayton or was your intent that the pod cache in the openshift-sdn-enabled proxy use the new caching pattern you're doing for DNS?  If so, I'd like to get the split proxy merged first and get people actually using it, and then optimize the pod watch if that's OK.

**Comment by smarterclayton** [comment]
My goal would be that we would not have a pod cache on the node.  I
want to have that discussion first and figure out what other bits of
the code have to be fixed.  We know we can move endpoints filtering
off node - getting pods off is the next step.

File: pkg/cmd/server/kubernetes/node_config.go [hunk 2]
```diff
@@ -300,19 +297,19 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
 	return config, nil
 }
 
-func buildKubeProxyConfig(options configapi.NodeConfig) (*proxyoptions.ProxyServerConfig, error) {
+func buildKubeProxyConfig(options configapi.NodeConfig, originClient *osclient.Client, kubeClient *client.Client) (*proxyoptions.ProxyServerConfig, osdnapi.FilteringEndpointsConfigHandler, error) {
```
Comment (danwinship) [inline pkg/cmd/server/kubernetes/node_config.go]: the filter isn't part of the proxy *config*... you could leave this as is and just add the call to `factory.NewProxyPlugin()` to `BuildKubernetesNodeConfig()`, right after the call to `factory.NewNodePlugin()`, and thereby simplify things a bunch...

File: pkg/cmd/server/start/node_args.go [hunk 3]
```diff
@@ -152,12 +152,6 @@ func ValidateRuntime(config *configapi.NodeConfig, components *utilflags.Compone
...
 	case ovs.MultiTenantPluginName:
-		if actual.Has(ComponentKubelet) && !actual.Has(ComponentPlugins) {
-			return fmt.Errorf("the multi-tenant SDN plugin must be run in the same process as the kubelet")
-		}
-		if actual.Has(ComponentProxy) && !actual.Has(ComponentPlugins) {
-			return fmt.Errorf("the multi-tenant SDN plugin requires the proxy and plugins components be enabled in the same process")
-		}
 	case ovs.SingleTenantPluginName:
 		if actual.Has(ComponentKubelet) && !actual.Has(ComponentPlugins) {
 			return fmt.Errorf("the SDN plugin must be run in the same process as the kubelet")
```
Comment (danwinship) [inline pkg/cmd/server/start/node_args.go]: You're leaving the "sdn in same process as kubelet" check for single-tenant, but not for multi-tenant

**Comment by danwinship** [comment]
needs a Godeps bump too now that the openshift-sdn party has merged

**Comment by danwinship** [comment]
> We know we can move endpoints filtering off node - getting pods off is the next step.

the pod cache is only used by the endpoint filtering, so once endpoint filtering is off-node there is no next step needed

**Comment by smarterclayton** [comment]
Ok, let's line up a brainstorm this week or next.  Securing endpoints is
useful in general  (much like externalIPs)

On May 4, 2016, at 4:18 PM, Dan Winship <notifications@github.com> wrote:

We know we can move endpoints filtering off node - getting pods off is the
next step.

the pod cache is only used by the endpoint filtering, so once endpoint
filtering is off-node there is no next step needed

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/8627#issuecomment-216989215>

**Comment by pravisankar** [comment]
On Wed, May 4, 2016 at 1:18 PM, Dan Winship <notifications@github.com>
wrote:

> We know we can move endpoints filtering off node - getting pods off is the
> next step.
>
> the pod cache is only used by the endpoint filtering, so once endpoint
> filtering is off-node there is no next step needed
>
No step required? yes, endpoint filtering off-node don't need pod cache in
the openshift SDN but we still need to solve the use case (services in one
namespace can't point to pods in another namespace) for multi-tenant.

**Comment by smarterclayton** [comment]
Yes, but the nodes is the wrong place to do it.

**Comment by danwinship** [comment]
> No step required? yes, endpoint filtering off-node don't need pod cache in the openshift SDN but we still need to solve the use case (services in one namespace can't point to pods in another namespace) for multi-tenant.

What I was trying to say is that "get rid of the pod cache" happens automatically as part of "move endpoint filtering off node". It's not a separate step that would still remain to be done after the filtering was moved.

**Comment by dcbw** [comment]
@danwinship update with fixes for your comments. PTAL

**Comment by danwinship** [comment]
[merge]

**Pull request merged**